mlx/backend/metal/compiled.cpp

```diff
@@ -1,44 +1,475 @@
 // Copyright © 2023-2024 Apple Inc.
 
+#include <sstream>
+
+#include "mlx/backend/metal/compiled_preamble.h"
 #include "mlx/backend/metal/device.h"
+#include "mlx/backend/metal/utils.h"
+#include "mlx/graph_utils.h"
 #include "mlx/primitives.h"
+#include "mlx/utils.h"
 
 namespace mlx::core {
 
-void Compiled::eval_gpu(
+inline bool is_static_cast(const Primitive& p) {
+  return (
+      typeid(p) == typeid(Broadcast) || typeid(p) == typeid(Copy) ||
+      typeid(p) == typeid(StopGradient) || typeid(p) == typeid(AsType));
+}
+
+inline auto get_type_string(Dtype d) {
+  switch (d) {
+    case float32:
+      return "float";
+    case float16:
+      return "half";
+    case bfloat16:
+      return "bfloat16_t";
+    case bool_:
+      return "bool";
+    case int8:
+      return "int8_t";
+    case int16:
+      return "int16_t";
+    case int32:
+      return "int32_t";
+    case int64:
+      return "int64_t";
+    case uint8:
+      return "uint8_t";
+    case uint16:
+      return "uint16_t";
+    case uint32:
+      return "uint32_t";
+    case uint64:
+      return "uint64_t";
+    default: {
+      std::ostringstream msg;
+      msg << "Unsupported compilation type " << d;
+      throw std::runtime_error(msg.str());
+    }
+  }
+}
+
+template <typename T>
+void print_float_constant(std::ostream& os, const array& x) {
+  auto old_precision = os.precision();
+  os << std::setprecision(std::numeric_limits<float>::digits10 + 1)
+     << x.item<T>() << std::setprecision(old_precision);
+}
```

Comment:
For metal `half` types, it's best to put `h` after literals (`2.0` vs `2.0h`) otherwise everything is promoted to `float`. That said I don't think that's a problem for the current paradigm. 
As far as I understand it, we aren't inlining any constants, in that it is always going to be 
`half a = ...; half b = 2.0; half c = Add()(a, b);` instead of `half a = ...; half c = Add()(a, 2.0);` so not much to worry about unless there is a place where we are inlining constants.

Reply:
Yeah exactly, I am also always printing them in float precision and they will automatically cast to whatever. Good to know for the future though!